web/src/inbox_ui.js

```diff
@@ -0,0 +1,875 @@
+        is_web_public: stream_info.is_web_public,
+        stream_name: stream_info.name,
+        stream_color: stream_color.get_stream_privacy_icon_color(stream_info.color),
+        stream_header_color: stream_color.get_recipient_bar_color(stream_info.color),
+        unread_count,
+        stream_url: hash_util.by_stream_url(stream_id),
+        stream_id,
+        // Will be displayed if any topic is visible.
+        is_hidden: true,
+    };
+}
+
+function validate_stream_data(new_stream_data, old_stream_data) {
+    for (const property in new_stream_data) {
+        if (new_stream_data[property] !== old_stream_data[property]) {
+            const $rendered_row = get_stream_header_row(new_stream_data.stream_id);
+            $rendered_row.replaceWith(render_inbox_row(new_stream_data));
+            return;
+        }
+    }
+}
+
+function format_topic(stream_id, topic) {
+    const topic_unread = unread.num_unread_for_topic(stream_id, topic);
+    const context = {
+        is_topic: true,
+        stream_id,
+        topic_name: topic,
+        unread_count: topic_unread,
+        conversation_key: get_topic_key(stream_id, topic),
+        topic_url: hash_util.by_stream_topic_url(stream_id, topic),
+        is_hidden: filter_should_hide_row({stream_id, topic}),
+    };
+
+    return context;
+}
+
+function insert_stream(stream_id, topic_dict) {
+    const stream_data = format_stream(stream_id);
+    const stream_key = get_stream_key(stream_id);
+    topics_dict[stream_key] = {};
+    for (const [topic, topic_unread_count] of topic_dict) {
+        const topic_key = get_topic_key(stream_id, topic);
+        if (topic_unread_count) {
+            const topic_data = format_topic(stream_id, topic);
+            topics_dict[stream_key][topic_key] = topic_data;
+            if (!topic_data.is_hidden) {
+                stream_data.is_hidden = false;
+            }
+        }
+    }
+
+    streams_dict[stream_key] = stream_data;
+
+    const sorted_stream_keys = get_sorted_stream_keys();
+    const stream_index = sorted_stream_keys.indexOf(stream_key);
+    const rendered_stream = render_inbox_stream_container({
+        topics_dict: {
+            [stream_key]: topics_dict[stream_key],
+        },
+        streams_dict,
+    });
+
+    if (stream_index === 0) {
+        $("#stream_container").prepend(rendered_stream);
+    } else {
+        const previous_stream_key = sorted_stream_keys[stream_index - 1];
+        $(rendered_stream).insertAfter(get_stream_container(previous_stream_key));
+    }
+    return get_stream_container(stream_key);
+}
+
+function validate_topic_data(new_topic_data, old_topic_data) {
+    // This row is not rendered yet.
+    if (old_topic_data === undefined) {
+        const stream_key = get_stream_key(new_topic_data.stream_id);
+        const $topic_container = get_stream_container(stream_key).find(".topic_contianer");
+        $topic_container.prepend(render_inbox_row(new_topic_data));
```

Comment:
Hmm, what's the logic behind prepending? Is the idea that this should only happen when new messages are coming in, and those are sorted to the top?

I'd worry a bit about sort order issues in the case that older topics were marked as unread via "mark as unread"

Reply:
> Is the idea that this should only happen when new messages are coming in, and those are sorted to the top?

Yeah.

> I'd worry a bit about sort order issues in the case that older topics were marked as unread via "mark as unread"

Yeah, I think some folks might prefer it that way. We can iterate on it later on.

Reply:
OK opened #26667 to track this.